### PR TITLE
[prebuilds] prebuild detail view in create workspace

### LINKED ISSUE #9132: Use the same prebuild view for Workspace start as for Prebuild Details
## Is your feature request related to a problem? Please describe

Originally raised in https://github.com/gitpod-io/gitpod/pull/8831#discussion_r842711075 by @gtsiolis.

| BEFORE | AFTER |
|-|-|
| Screenshot 2022-04-05 at 2 00 24 PM | Screenshot 2022-04-05 at 2 09 51 PM |

Doing this would normalize the UI elements between prebuilds and workspace starts (which require a prebuild)

### PR DESCRIPTION
**Pull request by laushinka**
## Description
This PR gives a perceived consistency between the create workspace page and the prebuild detail view.  

Screenshot 2022-06-22 at 00 36 13

## Related Issue(s)
Fixes #9132

## How to test
1. Make a commit to a project with a prebuild.
2. Go to the project page's Branches tab. Once the prebuild is shown as "Running", start a new workspace.
3. See the logs running and the prebuild status showing below.

## Release Notes

```release-note
Prebuild status is shown under the logs when starting a workspace.
```

## Documentation

- [x] /werft with-preview
- [x] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by roboquat** [comment]
@laushinka: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/dashboard/src/start/IntermediatePrebuildStatus.tsx
Comment (laushinka) [inline components/dashboard/src/start/IntermediatePrebuildStatus.tsx:9]: To show the prebuild status, we need to find prebuilds related to a project. When showing the logs in the prebuild detail view, we already have the project information. This is not the case during workspace start. However, we do have a prebuild ID and status. 
This intermediate status component is a workaround so that we can show the prebuild status "Running" while waiting for the prebuild updates. Once the prebuild updates are received and we have a prebuild object, we show those accordingly.

Comment (laushinka) [reply components/dashboard/src/start/IntermediatePrebuildStatus.tsx]: We don't need this intermediate status anymore because we are now fetching the prebuild from the beginning.

**Comment by jldec** [comment]
From my experiments in this preview environment, the scrollbar in the project prebuild log view (1st screenshot below) is quite different from the scrollbar in the start workspace view with prebuild in progress. (2nd screenshot below)
 
Screenshot 2022-06-23 at 12 03 38

Screenshot 2022-06-23 at 12 04 06

**Comment by jldec** [comment]
Other than the scrollbar difference, the behavior LGTM
/hold for more technical review

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (geropl) [inline components/dashboard/src/components/PrebuildLogs.tsx:26]: This one looks like a copy'n'paste thing; it's not used (anymore).

Comment (laushinka) [reply components/dashboard/src/components/PrebuildLogs.tsx:26]: Thanks for the catch! Will remove it.

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (geropl) [inline components/dashboard/src/components/PrebuildLogs.tsx:51]: We should always be able to retrieve the prebuild shape. For that we could use a method called `findPrebuildByWorkspaceId(workspaceId)`.

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (geropl) [inline components/dashboard/src/components/PrebuildLogs.tsx:70]: This will get triggered by _any_ prebuild this user has access to.
Maybe a simple `if (update.info && update.info.buildWorkspaceId === props.workspaceId)` already is enough.

File: components/dashboard/src/projects/Prebuild.tsx
Comment (geropl) [inline components/dashboard/src/projects/Prebuild.tsx:159]: :orange_heart:

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [inline components/dashboard/src/start/CreateWorkspace.tsx:497]: Why change the text here?
Personally I don't have an opinion, but these texts tend to have been checked by other people having a better overview over UX then myself. :wink:

File: components/dashboard/src/start/StartWorkspace.tsx
Comment (geropl) [inline components/dashboard/src/start/StartWorkspace.tsx:395]: Should be: `const isPrebuild = this.state.workspace?.type === "prebuild"`

File: components/dashboard/src/start/StartWorkspace.tsx
Comment (geropl) [inline components/dashboard/src/start/StartWorkspace.tsx:396]: :100:

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (laushinka) [reply components/dashboard/src/start/CreateWorkspace.tsx:497]: Good point! I just followed the text specified in the issue because I assumed that went through a design discussion already. Thoughts, @jldec?

File: components/server/src/projects/projects-service.ts
Comment (geropl) [inline components/server/src/projects/projects-service.ts:228]: This should not reside in ProjectService I feel. It has no connection to that, at all. I'd rather see it in GitpodServerImpl - or a PrebuildsService, if we were to introduce one.

File: components/dashboard/src/start/StartWorkspace.tsx
Comment (laushinka) [reply components/dashboard/src/start/StartWorkspace.tsx:734]: Oh right, that's a good point. Now everything it needs is handled already in PrebuildLogs.

File: components/server/src/projects/projects-service.ts
Comment (laushinka) [reply components/server/src/projects/projects-service.ts:228]: My thought was because prebuilds are intended to only be used for projects, therefore I kept it here. But I can see why it seems out of place. I'll move it 👍🏼

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (laushinka) [reply components/dashboard/src/components/PrebuildLogs.tsx:51]: Currently the only PrebuildLogs needing a prebuild are CreateWorkspace and Prebuild, and both have this information already and it gets passed into the component as a prop. Do we need to make this extra call?

Comment (laushinka) [reply components/dashboard/src/components/PrebuildLogs.tsx:51]: Oh wait StartWorkspace also

Comment (geropl) [reply components/dashboard/src/components/PrebuildLogs.tsx:51]: > Do we need to make this extra call?

It's not an extra call. So far we doing 1 call (getPrebuild) sometimes, and I'm arguing for doing 1 call (findPrebuildByWorkspaceId) always.
If we wanted to optimize we could think about accepting a full `PrebuildInfo` shape to avoid rendering placeholders but we have to do that one call anyway to be sure we're up-to-date and avoid lost updates.

File: components/dashboard/src/start/CreateWorkspace.tsx
Comment (geropl) [reply components/dashboard/src/start/CreateWorkspace.tsx:497]: :bulb: Ah, didn't noticed!

File: components/server/src/projects/projects-service.ts
Comment (geropl) [reply components/server/src/projects/projects-service.ts:228]: > I'll move it

:pray:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:2248]: One last - but important - thing: We need to guard access to a prebuild.

So far, we don't have an access guard for prebuilds specifically (I think we should have one).
As placeholder, we mostly used the [WorkspaceInstance one](https://github.com/gitpod-io/gitpod/blob/3389f9a91c2e9401c2e49d97cf5b7d241d4a37bf/components/server/src/workspace/gitpod-server-impl.ts#L1542-L1543), which is ok to re-use in the scope of this PR.

But we should create a follow-up issue to fix this. :+1:

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (laushinka) [reply components/dashboard/src/components/PrebuildLogs.tsx]: Incorporated `findPrebuildByWorkspaceId` and got rid of the `prebuildId` props 👍🏼

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:2249]: Thanks for adding this one! :pray:
Instead of `trace({})` we should forward the `ctx` I think.

File: components/dashboard/src/start/StartWorkspace.tsx
Comment (geropl) [reply components/dashboard/src/start/StartWorkspace.tsx:734]: `HeadlessWorkspaceView` can be replace entirely I think. :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:2249]: Ah, and we still need some form of access restriction, as mentioned [here](https://github.com/gitpod-io/gitpod/pull/10696#discussion_r904965697).

Copying over the lines from the headless-log method should be fine. :+1:

File: components/dashboard/src/start/StartWorkspace.tsx
Comment (laushinka) [reply components/dashboard/src/start/StartWorkspace.tsx:734]: Ahhh I already removed this but then put it back for testing and mistakenly committed it

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (laushinka) [reply components/server/src/workspace/gitpod-server-impl.ts]: @geropl The existing access guards didn't work because they check for properties that the prebuildWithStatus didn't have, so I created a basic `GuardedPrebuild` resource. Tested the preview env and seems to work fine, but let me know what you think 🙏🏽

Comment (laushinka) [reply components/server/src/workspace/gitpod-server-impl.ts]: Removed ✅

**Comment by geropl** [comment]
This is basically done, and I'd love to approve. But as [we found out above](https://github.com/gitpod-io/gitpod/pull/10696#discussion_r906208239), we're missing proper resourceguards for prebuilds. So I'm going to hold this PR, and try to get it in before we land this.

/hold

**Comment by geropl** [comment]
Now waiting for https://github.com/gitpod-io/gitpod/pull/10940 to get merged...

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts]: Fixed with https://github.com/gitpod-io/gitpod/pull/10696/commits/8f2b6c8be82dcff9c63830ef440e4f61f50464fc

**Review by gtsiolis — COMMENTED** [review]
I never clicked submit but here's what I had in draft since yesterday. 🙈 

Since we do not have yet the capability to expand the logs into a full page component, going with the current logs width sounds ok. Any thoughts on https://github.com/gitpod-io/gitpod/issues/9132#issuecomment-1092072758?

Two issues that pop up:
1. The scrollbar position mentioned in https://github.com/gitpod-io/gitpod/pull/10696#issuecomment-1164275542. Moving forward with the current scrollbar styles sounds ok but the positioning the scrollbar in the middle of the element feels like something is broken.
2. The width of the element wrapping the bottom parts of the prebuilds logs differ per case. Could we adjust the width of the bottom elements to match the prebuilds logs element above? See screenshots below.

| Prebuild (Running) | Prebuild (Preparing) |
|-|-|
| in-progress-running | in-progress-preparing |

**Review by geropl — APPROVED** [review]
Tested and works! :+1: 

/hold in cases someone else wants to review

**Review by gtsiolis — APPROVED** [review]
Perfection achieved. ✨ 

Thanks, @geropl, @laushinka, @jankeromnes, and @jldec! 🏀

While I still think using a narrower width as described in https://github.com/gitpod-io/gitpod/issues/9132#issuecomment-1092072758 could be interesting to explore, :a: the changes here already significantly improve the current state of the prebuild-in-progress page and :b: we'd still need to add a function to expand the view to a full page component which would significantly increase the scope of this PR.

Looking forward for https://github.com/gitpod-io/gitpod/issues/10041 or whatever comes next.

**Pull request merged**
